### PR TITLE
feat(ws1): Add PyTorch RoPE reference operator

### PR DESCRIPTION
**Pull request by frank-2077**
## Summary

Implements `NativeRoPEOp`, the fp32 ground-truth RoPE reference operator for ISSUE #108.

- **Shapes**: Qwen3-8B defaults — `theta=1e6` (not the 1e4 default), `head_dim=128`, full-dimension rotation (`half=64`).
- **Dtype contract**: `forward()` follows input dtype; `forward_fp32()` is the fp32 gold standard with internal fp32 accumulation. Both implemented (fp16 and fp32 paths tested).
- **Purity**: no in-place ops, no random state, no global state. cos/sin are computed internally in fp32 from `positions`/`theta` — no external cache accepted or returned.

## Changes

- `rl_engine/kernels/ops/pytorch/pos/rope.py` — `NativeRoPEOp` (`op_class = "elementwise"`)
- `docs/operators/rope.md` — operator reference doc (shapes, convention, tolerance contract)
- `tests/ops/test_rope.py` — full test suite, all green locally:
  - **Axis A (bitwise batch invariance)**: `torch.equal` between batch=1 slice and batch=N slice, including padded batches and per-row 2D `positions` (`[B, S]`)
  - **Axis B (accuracy)**: `forward` vs `forward_fp32` under tolerance contract (fp32: 1e-5, bf16: 2e-2, fp16: 1e-3)
  - Bitwise match against the frozen `ISSUE_108_OPS_DEV.md` §5 reference formula
  - Shape/dtype correctness, position broadcasting equivalence (`[S]` vs `[B,S]`), pure-function checks
  - Qwen3-8B specific shapes (Q heads=32, KV heads=8, head_dim=128)

image

## Summary by CodeRabbit

* **New Features**
  * Added a RoPE (Rotary Position Embeddings) operator with a PyTorch-native backend for CPU, CUDA, and ROCm.
* **Documentation**
  * Updated the operator index and added detailed RoPE operator docs, including expected tensor inputs, reference behavior, and test coverage.
* **Tests**
  * Added a comprehensive RoPE test suite validating exact fp32 correctness vs an independent reference, dtype/shape handling, position semantics, theta sensitivity, and batch invariance.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
please resolve conflicts and CI error first, Thanks.

File: rl_engine/kernels/ops/pytorch/rotary_embedding/rope.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        half = D // 2
+
+        # inv_freq[i] = 1 / (theta^(2i/D)) = 1 / (theta^(i/half))
+        # shape: [half]
+        inv_freq = 1.0 / (
+            theta ** (torch.arange(0, half, dtype=torch.float32, device=x.device) / half)
+        )
+
+        # positions: [S] -> [S, 1] or [B, S] -> [B, S, 1]
+        pos_float = positions.float().unsqueeze(-1)
```
Comment (KJLdefeated) [inline rl_engine/kernels/ops/pytorch/rotary_embedding/rope.py]: The positions may not have same device with inv_freq. Maybe you should convert it to device=x.device

**Review by KJLdefeated — CHANGES_REQUESTED** [review]
This PR looks good to me. After resolve the requests, I am happy to approve.

File: tests/test_rope.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        op = NativeRoPEOp()
+        x, pos = _make_inputs(2, 32, 16, QWEN3_HEAD_DIM)
+        x_typed = x.to(dtype)
+        out_typed = op.forward(x_typed, pos).float()
+        out_fp32 = op.forward_fp32(x_typed, pos)
+        diff = (out_typed - out_fp32).abs().max().item()
+        assert torch.allclose(out_typed, out_fp32, atol=atol, rtol=rtol), (
+            f"dtype={dtype}, max_abs_error={diff:.3e} exceeds " f"atol={atol}, rtol={rtol}"
+        )
+
```
Comment (Flink-ddd) [inline tests/test_rope.py:235]: Unlike the RMSNorm and SwiGLU PRs, this PR currently has zero tests for backward(). Even though RoPE has no trainable weights, gradients must flow back through it accurately to the Query and Key projection layers.

You need to verify that x.grad is correctly computed and is strictly Batch-Invariant (Axis-A), otherwise upstream training drift is guaranteed. Please add a slice test to assert that computing the backward pass on a full batch yields bitwise identical gradients to computing it on a sliced subset (batch size = 1).

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Focusing strictly on the remaining architectural blind spots to ensure this PR meets the exact same WS1 standards as #160 and #166, here is the review:

**Comment by Flink-ddd** [comment]
Please resolve the code conflicts first, Thanks.

**Comment by Flink-ddd** [comment]
Please resolve the code conflicts first. If the review information has already resolved the issue or only requires explanation, please mark it. If there are no other issues, we will merge this PR. Thank you.

File: rl_engine/kernels/ops/pytorch/rotary_embedding/rope.py (hunk 1 above)
Comment (frank-2077) [reply rl_engine/kernels/ops/pytorch/rotary_embedding/rope.py]: Fixed. positions are now moved to x.device before computing freqs, so positions and inv_freq are always on the same device.

File: tests/test_rope.py (hunk 2 above)
Comment (frank-2077) [reply tests/test_rope.py:235]: Added a backward slice test for RoPE. It verifies that x.grad from the full batch is bitwise identical to x.grad computed from each batch=1 slice.

**Review by Flink-ddd — APPROVED** [review]
LGTM, Thanks. cc @KJLdefeated PTAL again.

**Review by KJLdefeated — APPROVED** [review]
LGTM, Thx!
